src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,98 @@
+
+
+class ModelManager(Generic[T]):
+    """Manage machine learning models with direct dictionary configuration."""
+
+    def __init__(self, model_config: dict[str, Path]) -> None:
+        """Initialize the model manager with a direct model configuration.
+
+        Args:
+            model_config: A dictionary mapping model keys to model file paths (as Path objects).
+        """
+        self._models: dict[str, _Model[T]] = {}
+        self.model_config = model_config
+        self._monitoring_task: asyncio.Task[None] | None = None
+        self.load_models()
+
+    def load_models(self) -> None:
+        """Load models based on the provided model configuration."""
+        for key, model_path in self.model_config.items():
+            self._models[key] = _Model(data=self._load(model_path), path=model_path)
+
+        # Start monitoring models for changes.
+        self._monitoring_task = asyncio.create_task(self._start_model_monitor())
+
+    def _load(self, model_path: Path) -> T:
+        """Load the model from the specified path."""
+        if not model_path.exists():
+            raise FileNotFoundError(f"The model path {model_path} does not exist.")
+
+        with model_path.open("rb") as file:
+            model = pickle.load(file)
+            return cast(T, model)
+
+    def get_model(self, key: str) -> T:
+        """Retrieve a loaded model by key."""
+        try:
+            return self._models[key].data
+        except KeyError as exc:
+            raise KeyError(f"Model with key '{key}' is not found.") from exc
+
+    async def _start_model_monitor(self) -> None:
+        """Start monitoring the model paths for changes."""
+        model_paths = [model.path for model in self._models.values()]
+        file_watcher = FileWatcher(paths=[str(path) for path in model_paths])
+
+        _logger.info("Monitoring the paths of the models for changes: %s", model_paths)
+
+        async for event in file_watcher:
+            if event.type in (EventType.CREATE, EventType.MODIFY):
+                _logger.info(
+                    "Model file %s has been modified. Reloading model...", event.path
+                )
+                self.reload_model(Path(event.path))
+
+    def reload_model(self, model_path: Path) -> None:
+        """Reload a model from the given path."""
+        for key, model in self._models.items():
+            if model.path == model_path:
+                model.data = self._load(model_path)
+                return
+        raise ValueError(f"No model found with path: {model_path}")
+
+    async def join(self) -> None:
+        """Await the monitoring task, and return when the task completes."""
+        if self._monitoring_task and not self._monitoring_task.done():
+            await self._monitoring_task
+
+    async def _stop_model_monitor(self) -> None:
+        """Stop monitoring the model paths for changes."""
+        if self._monitoring_task:
+            await cancel_and_await(self._monitoring_task)
```

Comment:
Oh, this makes me think this class should be a [`BackgroundService`](https://frequenz-floss.github.io/frequenz-sdk-python/v0.25/reference/frequenz/sdk/actor/#frequenz.sdk.actor.BackgroundService) instead, then you will get the `join()` and `stop()` methods automatically, and we keep a consistent interface.

Reply:
Just to clarify we wouldn't have `join()` right? We would have `wait()` instead?

Reply:
Just to clarify we wouldn't have `join()` right? We would use `wait()` instead?

Reply:
Ok I just read a bit more on `BackgroundService`. We wouldn't even need to implement it at all I see since it is already taken care of.

Reply:
Done!